File: src/houseinfo.rs
use anyhow::{bail, Context, Ok, Result};
use serde::Deserialize;
use std::collections::BTreeMap;
use std::collections::HashMap;
use std::path::Path;

pub(crate) type IndexedStringMap = HashMap<usize, String>;
pub(crate) type BuildingMap = BTreeMap<usize, Building>;
pub(crate) type RegionBuildingMap = BTreeMap<String, BuildingMap>;
pub(crate) type CraftBuildingMap = BTreeMap<String, (usize, Vec<Building>)>;

// pub const HOUSECRAFT_TABLE_STYLE: &str = "0123456789abcdefghi";
pub(crate) const HOUSECRAFT_TABLE_STYLE: &str = "   ═────      ═  ══";

#[derive(Clone, Debug)]
pub(crate) struct UsageCounters {
    pub cost: usize,
    pub warehouse_count: usize,
    pub worker_count: usize,
}

impl UsageCounters {
    pub fn new() -> Self {
        Self {
            cost: 0,
            warehouse_count: 0,
            worker_count: 0,
        }
    }
}

#[allow(unused)]
#[derive(Clone, Debug)]
pub(crate) struct Building {
    pub key: usize,
    pub need_key: usize,
    pub node_key: usize,
    pub region_key: usize,
    pub building_name: String,
    pub node_name: String,
    pub region_name: String,
    pub cost: usize,
    pub worker_count: usize,
    pub stable_count: usize,
    pub warehouse_count: usize,
    pub craft_list: Vec<CraftList>,
}

impl Building {
    pub fn new(
        house_info: &HouseInfo,
        character: &IndexedStringMap,
        exploration: &IndexedStringMap,
        region: &IndexedStringMap,
    ) -> Self {
        let key = house_info.character_key;
        let node_key = house_info.parent_node;
        let region_key = house_info.affiliated_town;

        Self {
            key,
            need_key: if house_info.len_need_house_key == 1 {
                house_info.need_house_key
            } else {
                house_info.affiliated_town
            },
            node_key,
            region_key,

            building_name: character.get(&key).unwrap().to_string(),
            node_name: exploration.get(&node_key).unwrap().to_string(),
            region_name: region.get(&region_key).unwrap().to_string(),

            cost: house_info.need_explore_point as usize,
            worker_count: house_info.worker_count(),
            warehouse_count: house_info.warehouse_count(),
            stable_count: house_info.stable_count(),

            craft_list: house_info.craft_list.clone(),
        }
    }
}
#[derive(Deserialize, Debug)]
pub(crate) struct HouseInfos(HashMap<String, HouseInfo>);

#[allow(unused)]
#[derive(Deserialize, Debug)]
pub(crate) struct HouseInfo {
    pub affiliated_town: usize,
    pub character_key: usize,       /* key [CharacterKey] */
    pub craft_list: Vec<CraftList>, /* house craft list [pa_vector<gc::HouseInfoCraft>] */
    pub len_need_house_key: usize,
    pub house_floor: usize,      /* house floor [HouseFloor] */
    pub house_group: usize,      /* house group [HouseGroup] */
    pub need_explore_point: u32, /* need explore count [ExplorationPoint] */
    pub need_house_key: usize,   /* need house key list [pa_vector<gc::CharacterKey>] */
    pub num_craft_list_items: u32,
    pub parent_node: usize, /* parent node key [gc::WaypointKey] */
}

impl HouseInfo {
    pub fn worker_count(&self) -> usize {
        self.craft_index_to_count(1)
    }

    pub fn warehouse_count(&self) -> usize {
        self.craft_index_to_count(2)
    }

    pub fn stable_count(&self) -> usize {
        self.craft_index_to_count(3)
    }

    fn craft_index_to_count(&self, index: usize) -> usize {
        self.craft_list
            .iter()
            .find(|c| c.item_craft_index == index)
            .map(|c| c.level_to_count() as usize)
            .unwrap_or(0)
    }
}

#[derive(Clone, Deserialize, Debug)]
pub(crate) struct CraftList {
    pub house_level: u32,
    pub item_craft_index: usize,
}

impl CraftList {
    pub fn level_to_count(&self) -> u32 {
        if self.item_craft_index == 0 || self.item_craft_index > 3 {
            return 0;
        }

        match self.item_craft_index {
            1 => match self.house_level {
                1 => 1,
                2 => 2,
                3 => 4,
                4 => 6,
                5 => 8,
                _ => 0,
            },
            2 => match self.house_level {
                1 => 3,
                2 => 5,
                3 => 8,
                4 => 12,
                5 => 16,
                _ => 0,
            },
            3 => match self.house_level {
                1 => 1,
                2 => 2,
                3 => 3,
                4 => 4,
                5 => 5,
                _ => 0,
            },
            _ => 0,
        }
    }
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "PascalCase")]
struct LanguageData {
    param0: usize,
    string: String,
}

pub(crate) fn read_csv_data(filename: &str) -> Result<IndexedStringMap> {
    let path = Path::new("./data/houseinfo/filename.txt");
    let path = path.with_file_name(filename);
    let path_string = path.clone();
    let path_string = path_string.as_path().display();
    let file = std::fs::File::open(path).context(format!("Can't find {path_string})"))?;
    let mut rdr = csv::Reader::from_reader(file);
    let mut records: std::collections::HashMap<usize, String> = std::collections::HashMap::new();
    for result in rdr.deserialize() {
        let record: LanguageData = result?;
        records.insert(record.param0, record.string);
    }
    Ok(records)
}

pub(crate) fn parse_houseinfo_data() -> Result<RegionBuildingMap> {
    let character = read_csv_data("Character.csv")?;
    let exploration = read_csv_data("Exploration.csv")?;
    let region = read_csv_data("Region.csv")?;

    let json_string = std::fs::read_to_string("./data/houseinfo/houseinfo.json")?;
    let house_infos_map: HashMap<String, HouseInfo> = serde_json::from_str(&json_string)?;
    let house_infos: Vec<HouseInfo> = house_infos_map.into_values().collect();

    let mut region_buildings = RegionBuildingMap::new();

    for house_info in house_infos.into_iter() {
        let building = Building::new(&house_info, &character, &exploration, &region);
        if let Some(region) = region_buildings.get_mut(&building.region_name) {
            region.insert(building.key, building);
        } else {
            let region_key = building.region_name.to_owned();
            let mut value = BuildingMap::new();
            value.insert(building.key, building);
            region_buildings.insert(region_key, value);
        }
    }
    Ok(region_buildings)
}

pub(crate) fn get_region_buildings(region_name: Option<String>) -> Result<RegionBuildingMap> {
    let mut region_buildings = parse_houseinfo_data()?;
    if let Some(region_name) = region_name {
        region_buildings.retain(|k, _| *k == region_name);
        if region_buildings.len() != 1 {
            bail!(
                "Unable to find region {}.\n Use '--list-regions'.",
                region_name
            );
        }
    };
    Ok(region_buildings)
}

File: python/houseinfo.py
"""houseinfo.json extraction/parsing functions."""

from csv import DictReader
import json


def get_affiliated_town_key(name):
    with open("./data/houseinfo/" + "regioninfo.json", "r") as f:
        regions = json.load(f)
    for k, v in regions.items():
        if v == name:
            return int(k)
    return 0


def get_worker_count(craft_list):
    """Return worker count from house level given crafting list."""
    level = 0
    for craft in craft_list:
        if craft.get("item_craft_index") == 1:
            level = craft.get("house_level")

    if level == 1:
        return 1
    if level == 2:
        return 2
    if level == 3:
        return 4
    if level == 4:
        return 6
    if level == 5:
        return 8
    return 0


def get_extend_warehouse_count(craft_list):
    """Return warehouse count from house level given crafting list."""
    level = 0
    for craft in craft_list:
        if craft.get("item_craft_index") == 2:
            level = craft.get("house_level")

    if level == 1:
        return 3
    if level == 2:
        return 5
    if level == 3:
        return 8
    if level == 4:
        return 12
    if level == 5:
        return 16
    return 0


def get_extend_stable_count(craft_list):
    """Return stable count from house level given crafting list."""
    level = 0
    for craft in craft_list:
        if craft.get("item_craft_index") == 3:
            level = craft.get("house_level")

    if level == 1:
        return 1
    if level == 2:
        return 2
    if level == 3:
        return 3
    if level == 4:
        return 4
    if level == 5:
        return 5
    return 0


def read_csv_data(filename):
    """Read csv data file from the ./data directory."""
    path = "./data/houseinfo/" + filename
    records = {}
    with open(path, encoding="UTF-8") as file:
        for record in DictReader(file):
            records[int(record["Param0"])] = record["String"]
    return records


def get_town_buildings_by_key(town_key):
    """Return all building for a town given the town id key."""
    character = read_csv_data("Character.csv")
    exploration = read_csv_data("Exploration.csv")
    region = read_csv_data("Region.csv")

    with open("./data/houseinfo/houseinfo.json", encoding="UTF-8") as file:
        house_infos = json.load(file)

    buildings = {}
    for house_info in house_infos.values():
        if house_info.get("affiliated_town") == town_key:
            key = house_info.get("character_key")
            need_key = (
                house_info.get("need_house_key")
                if house_info.get("len_need_house_key") == 1
                else house_info.get("affiliated_town")
            )
            node_key = house_info.get("parent_node")
            region_key = house_info.get("affiliated_town")

            building_name = character.get(key)
            node_name = exploration.get(node_key)
            region_name = region.get(region_key)

            cost = house_info.get("need_explore_point")
            worker_count = get_worker_count(house_info.get("craft_list"))
            warehouse_count = get_extend_warehouse_count(house_info.get("craft_list"))
            stable_count = get_extend_stable_count(house_info.get("craft_list"))

            building = {
                "key": key,
                "need_key": need_key,
                "node_key": node_key,
                "region_key": region_key,
                "building_name": building_name,
                "node_name": node_name,
                "region_name": region_name,
                "cost": cost,
                "worker_count": worker_count,
                "stable_count": stable_count,
                "warehouse_count": warehouse_count,
            }
            buildings[key] = building
    return buildings


def get_town_buildings_by_name(town_name):
    """Return all buildings for a town given the town name."""
    town_key = get_affiliated_town_key(town_name)
    return get_town_buildings_by_key(town_key)


def get_region_buildings(region_name):
    """Prepare town buildings as a town hierarchy with parents, children, states and values
    for both warehouse and worker slots.
    """
    town_key = get_affiliated_town_key(region_name)
    if town_key == 0:
        raise SystemExit(f"{region_name} was not found.")
    buildings = get_town_buildings_by_name(region_name)
    keys = [town_key]
    needed_keys = [0]
    costs = [0]
    state_1_values = [0]
    state_2_values = [0]
    for key, value in buildings.items():
        keys.append(key)
        needed_keys.append(value.get("need_key"))
        costs.append(value.get("cost"))
        state_1_values.append(value.get("warehouse_count"))
        state_2_values.append(value.get("worker_count"))

    return buildings, keys, needed_keys, costs, state_1_values, state_2_values
